### PR TITLE
Transformers: Support inner vs outer join

### PR DESCRIPTION
**Pull request by ryantxu**
This extends the work from @AlexKaracaoglu in https://github.com/grafana/grafana/pull/53865 and adds it to the existing transformer:

Is this a breaking change?  I don't think so -- the old name "seriesToFields" will still work, but we will internally structure things to be based on the join name

TODO:
- [x] refactor `SeriesToFieldsTransformer` to `JoinTransformer`
- [x] update docs
- [x] add devenv dashboard with examples

### CODE REVIEW

**Comment by ryantxu** [comment]
I also tried a radio... not as good because I think we need to more clearly define the difference between INNER+OUTER:

image

File: docs/sources/panels/transform-data/index.md [hunk 1]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+Result after applying the inner join transformation:
+
+| Time                | Job     | Uptime    | Server   | Errors |
+| ------------------- | ------- | --------- | -------- | ------ |
+| 2020-07-07 11:34:20 | node    | 25260122  | server 1 | 15     |
+| 2020-07-07 11:24:20 | postgre | 123001233 | server 2 | 5      |
+
+#### Outer join
+
+An outer join will include all data from an inner join, but will also include rows where values do not match in every input.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:366]:
```suggestion
An outer join includes all data from an inner join and rows where values do not match in every input.
```

File: docs/sources/panels/transform-data/index.md [hunk 2]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+
+Use this transformation to combine the results from multiple queries (combining on a passed join field or the first time column) into one single result and drop rows where a successful join isn't able to occur - performing an inner join.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:368]:
```suggestion
Use this transformation to combine the results from multiple queries (combining on a passed join field or the first time column) into one result, and drop rows where a successful join cannot occur - performing an inner join.
```

Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:368]: I'm not sure what the last bit means " - performing an inner join." Should this read, ...where a successful join cannot occur, which perform an inner join"?

File: docs/sources/panels/transform-data/index.md [hunk 3]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+
+In the example below, we have two queries returning table data. It is visualized as two separate tables before applying the inner join transformation.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:370]:
```suggestion
In the following example, two queries return table data. It is visualized as two tables before applying the inner join transformation.
```

File: docs/sources/panels/transform-data/index.md [hunk 4]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+| 2020-07-07 11:04:20 | server 3 | 10     |
+
+Result after applying the inner join transformation:
+
+| Time                | Job     | Uptime    | Server   | Errors |
+| ------------------- | ------- | --------- | -------- | ------ |
+| 2020-07-07 11:34:20 | node    | 25260122  | server 1 | 15     |
+| 2020-07-07 11:24:20 | postgre | 123001233 | server 2 | 5      |
 
 In the example below, I have a template query displaying time series data from multiple servers in a table visualization. I can only view the results of one query at a time.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:395]:
```suggestion
In the following example, a template query displays time series data from multiple servers in a table visualization. The results of only one query can be viewed at a time.
```

File: docs/sources/panels/transform-data/index.md [hunk 5]
```diff
@@ -324,11 +324,73 @@ We would then get :
 
 This transformation allows you to extract some key information out of your time series and display them in a convenient way.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:325]:
```suggestion
This transformation enables you to extract key information from your time series and display it in a convenient way.
```

File: docs/sources/panels/transform-data/index.md [hunk 6]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+### Join by field
 
-Use this transformation to join multiple time series from a result set by field.
+Use this transformation to join multiple results into a single table. This is especially useful for converting multiple
+time series results into a single wide table with a shared time field.
 
-This transformation is especially useful if you want to combine queries so that you can calculate results from the fields.
+#### Inner join
+
+An inner join will merge data from multiple tables where all tables share the same value from the selected field. This will exclude
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:334]:
```suggestion
An inner join merges data from multiple tables where all tables share the same value from the selected field. This type of join excludes
```

File: docs/sources/panels/transform-data/index.md [hunk 7]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+data where values do not match in every result.
+
+Use this transformation to combine the results from multiple queries (combining on a passed join field or the first time column) into one single result and drop rows where a successful join isn't able to occur - performing an inner join.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:337]:
```suggestion
Use this transformation to combine the results from multiple queries (combining on a passed join field or the first time column) into one result, and drop rows where a successful join cannot occur - performing an inner join.
```

File: docs/sources/panels/transform-data/index.md [hunk 8]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+
+In the example below, we have two queries returning table data. It is visualized as two separate tables before applying the inner join transformation.
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:339]:
```suggestion
In the following example, two queries return table data. It is visualized as two separate tables before applying the inner join transformation.
```

File: docs/sources/panels/transform-data/index.md [hunk 9]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
+
+Query B:
+
+| Time                | Server   | Errors |
+| ------------------- | -------- | ------ |
+| 2020-07-07 11:34:20 | server 1 | 15     |
+| 2020-07-07 11:24:20 | server 2 | 5      |
+| 2020-07-07 11:04:20 | server 3 | 10     |
+
+Result after applying the inner join transformation:
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:357]:
```suggestion
The result after applying the inner join transformation looks like the following:
```

File: docs/sources/panels/transform-data/index.md [hunk 10]
```diff
@@ -324,11 +324,73 @@ We would then get :
...
```
Comment (chri2547) [inline docs/sources/panels/transform-data/index.md:388]:
```suggestion
The result after applying the inner join transformation looks like the following:
```

**Review by chri2547 — COMMENTED** [review]
Some suggestions and a question. Thanks!

File: docs/sources/panels/transform-data/index.md (hunk 3 above)
Comment (nmarrs) [reply docs/sources/panels/transform-data/index.md]: Isn't this section about applying an `outer join` transformation vs `inner`?

File: packages/grafana-data/src/transformations/transformers.ts [hunk 11]
```diff
@@ -34,7 +34,9 @@ export const standardTransformers = {
   reduceTransformer,
   concatenateTransformer,
   calculateFieldTransformer,
-  seriesToColumnsTransformer,
+  joinByFieldTransformer,
+  /** @deprecated */
+  seriesToColumnsTransformer: joinByFieldTransformer,
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/transformers.ts:39]: If I am understanding this correctly we are removing the old `seriesToColumnsTransformer` implementation and stubbing it with `joinByFieldTransformer` implementation. To allow for backwards compatibility we are deprecating the original option to eventually remove it in a future version in favor of keeping just `joinByFieldTransformer` correct?

File: packages/grafana-data/src/transformations/transformers/joinDataFrames.ts [hunk 12]
```diff
@@ -34,7 +34,7 @@ export function pickBestJoinField(data: DataFrame[]): FieldMatcher {
 }
 
 /**
- * @alpha
+ * @internal
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/transformers/joinDataFrames.ts:37]: Do we have documentation regarding these tags (i.e. definition / when to use them?) - in this case does marking this interface as `internal` vs `alpha` imply that users of @grafana/data shouldn't import this interface / use it in their own code? Do these tags trigger warnings if that is the case?

**Review by nmarrs — APPROVED** [review]
Tested out locally via gdev dashboards and some dummy data and works beautifully!

Had a few general questions / clarifications, also a nit with documentation where we seem to be referring to performing an inner join in the outer join section

File: packages/grafana-data/src/transformations/transformers.ts (hunk 11 above)
Comment (ryantxu) [reply packages/grafana-data/src/transformations/transformers.ts:39]: yup -- the name `seriesToColumns` is so old and does not really make sense anymore.  This has been called "Join (outer)" or "timeseries join" for so long that it is confusing to find it by the internal name "seriesToColumns" (a name that make sense in grafana 6)

File: docs/sources/panels/transform-data/index.md (hunk 2 above)
Comment (ryantxu) [reply docs/sources/panels/transform-data/index.md]: I fixed it :)

**Pull request merged**